src/pip/_internal/metadata/_json.py

```diff
@@ -64,10 +64,11 @@ def sanitise_header(h: Union[Header, str]) -> str:
         key = json_name(field)
         if multi:
             value: Union[str, List[str]] = [
-                sanitise_header(v) for v in msg.get_all(field)
+                sanitise_header(v)
+                for v in msg.get_all(field)  # type: ignore[union-attr]
             ]
         else:
-            value = sanitise_header(msg.get(field))
+            value = sanitise_header(msg.get(field))  # type: ignore[arg-type]
```

Comment:
What changed here?

Reply:
I'm guessing something now thinks `get` can return None, because it can't infer from the earlier `if field not in msg` that the field is guaranteed to be present.

I think this sort of ignore needs a more extensive comment, as it feels very much like a workaround for a limitation in mypy's ability to infer types correctly, or a lack of expressiveness in the type system to annotate `get` in such a way that this sort of LBYL idiom is supported.

We could of course add an assert stating that the return from get is not None, but I think that sort of assert is counterproductive, and obscures the actual logic here.

Reply:
Yup, I mentioned this in the body of the PR: https://github.com/python/typeshed/pull/9620
I can add a comment, should I do so in all affected places or only here?

Reply:
I don't honestly know. This is the sort of "typing is frustrating" issue I've been going on about on Discourse. I dislike `#type: ignore`  because it normalises switching off typing without documenting *why*. After all, someone could remove the check and then we're ignoring a genuine error.

I guess we should add a comment like "field is guaranteed to be present (see check on line XXX)". Unfortunately, the `#type: ignore` syntax doesn't appear to allow explanatory trailing text, which would be preferable.

At the end of the day, I'm just expressing my frustration. I don't really mind how we address this, it's not like we're worse off than we would be without type annotations at all.

Reply:
For this particular one I’d just add an explanation comment here. Other occurrences are in tests and those are less problematic especially a reader can more easily find the comment by tracing the test code.